README.md
[![API](https://img.shields.io/badge/API-16%2B-blue.svg?style=flat)](https://android-arsenal.com/api?level=16)
[![](https://jitpack.io/v/samani/Preflin.svg)](https://jitpack.io/#samani/Preflin)
# Preflin

Preflin is a wrapper over the SharedPreference class in Android. It supports storing objects and list objects other than the standard primitives while decreasing the boiler plate code. It also enables listening for changes on specific keys on a SharedPreference with RxJava support.

## Advantages

* Store all primitives include `double`, `float`, `string` and `integer`
* Store complex data objects and list objects
* Support RxJava change listener
* Support Jackson for serialize and deserialize objects

How to Use
-------


#### Gradle Configuration

Add jitpack.io to your root gradle file (project level) :
```gradle
allprojects {
  		repositories {
  			...
  			maven { url 'https://jitpack.io' }
  		}
 }
```

Add the dependency in your app build.gradle
```gradle
dependencies {
    implementation 'com.github.samani.Preflin:preflin:1.1.3'
    implementation 'com.github.samani.Preflin:preflin-gson:1.1.3' // Serialize object with Gson
    implementation 'com.github.samani.Preflin:preflin-jackson:1.1.3'  // Serialize object with Jackson
}
```

#### Usage

First initialize the Preflin library in your Application class.

```java
public class Application extends Application {
    @Override
    public void onCreate() {
        super.onCreate();
        Preflin.init(this);
        // Preflin With Gson Serializer
        Preflin.init(this, new GsonSerializer());
        // Preflin With Jackson Serializer
        Preflin.init(this, new JacksonSerializer());
    }
}
```

Then, use the Preflin class to store and access data in SharedPreferences.

```java
public class MainActivity extends Activity {

    @Override
    public void onCreate(Bundle inState) {
        // Default SharedPreferences
        Preflin.putString("title", "Deep Learning");
        Preflin.putDouble("page", 300);
        Preflin.putObject("book", new Book("Java", 300));
        Preflin.putObjectList("book_list", Arrays.asList(new Book("Java", 250),new Book("C++", 400)));

        Double page = Preflin.getDouble("page");
        Book javaBook = Preflin.getObject("book", Book.class);
        List<Book> bookList = Preflin.getObjectList(BOOK_LIST, Book.class);

        // Named SharedPreferences can also be used
        Preflin.from("Scores").putIntList("level_1", [60, 20, 80, 25, 30]);

        List<Integer> scores = Preflin.from("Scores").getIntList("level_1");
        
        //Check Key exists or not
        boolean isExists = Preflin.isKeyExists("book");
        
        //Remove element by Key
        Preflin.deleteValue("book");
        
        //Clear session
        Preflin.clearSession();
    }
}
```

Listen on changes to specific key in a SharedPreference

```java
// Listen on a key in default SharedPreferences
Preflin.listenOn("page").subscribe(new Action1<String>() {
    @Override
    public void call(String key) {
        Log.d(TAG, "1. Value of " + key + " changed to " + Preflin.getInt(key));
    }
});

// Listen on a key in named SharedPreferences
Preflin.from("scores").listenOn("user1").subscribe(new Action1<String>() {
    @Override
    public void call(String key) {
        Log.d(TAG, "2. Scores of " + key + " changed to " + Preflin.from(SP_SCORES).getDoubleList(key).toString());
    }
});
```

Do note that these subscriptions are standard RxJava subscriptions, hence they need to be unsubscribed, failing which will lead to leaking of subscriptions

Create Setter And getter and constructor for jackson

## Roadmap

1. ~~Gson support~~
2. Encryption support
3. ~~Multi module support (core, jackson, gson)~~ Publish on jitpack

License
-------

    Copyright 2018 - 2019 Rasoul Samani

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.


CHANGELOG.md
Change Log
==========

Version 1.1.3 *(2019-01-29)*
----------------------------

Publish modules on Jitpack.

Version 1.1.0 *(2019-01-28)*
----------------------------

 * Add `JacksonSerializer` and `GsonSerializer` for serialize and deserialize object
 * Add `clearSession()`.
 
Version 1.0.0 *(2019-01-25)*
----------------------------

Initial release.

app/src/main/java/com/rsamani/preflin/demo/MainActivity.java
package com.rsamani.preflin.demo;

import android.os.Bundle;
import android.os.Handler;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import com.rsamani.preflin.Preflin;

public class MainActivity extends AppCompatActivity {

    public static final String SP_SCORES = "SP_SCORES";
    public static final String PAGE = "PAGE";
    public static final String USER1 = "USER1";
    public static final String BOOK1 = "BOOK1";
    public static final String BOOK_LIST = "BOOK_LIST";
    private static final String TAG = "MainActivity";

    private CompositeDisposable disposables;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        disposables = new CompositeDisposable();

        // Named SharedPreferences
        Preflin.from(SP_SCORES).putDoubleList(USER1, Arrays.asList(1.1, 2.2, 3.33));
        Log.d(TAG, "scores of USER1 =" + Preflin.from(SP_SCORES).getDoubleList(USER1).toString());

        // Default SharedPreferences
        Preflin.putObject(BOOK1, new Book("Deep Learning", 400));
        Log.d(TAG, "BOOK1=" + Preflin.getObject(BOOK1, Book.class));

        List<Book> bookList = new ArrayList<>();
        bookList.add(new Book("Java", 250));
        bookList.add(new Book("C++", 400));
        Preflin.putObjectList(BOOK_LIST, bookList);
        Log.d(TAG, "BOOK_LIST=" + Preflin.getObjectList(BOOK_LIST, Book.class));

        Button incrementButton = (Button) findViewById(R.id.increment);
        incrementButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Log.d(TAG, "Incrementing page from " + Preflin.getInt(PAGE, 0));
                Preflin.putInt(PAGE, Preflin.getInt(PAGE, 0) + 1);
            }
        });

        // Listen on a key for default SharedPreferences
        Disposable disposable1 = Preflin.listenOn(PAGE).subscribe(new Consumer<String>() {
            @Override
            public void accept(String key) throws Exception {
                Log.d(TAG, "1. Value of " + key + " changed to " + Preflin.getInt(key));
                Toast.makeText(getApplicationContext(),
                        "1. Value of " + key + " changed to " + Preflin.getInt(key), Toast.LENGTH_SHORT).show();
            }
        });
        disposables.add(disposable1);

        // Listen on a key for named SharedPreferences
        Disposable disposable2 = Preflin.from(SP_SCORES).listenOn(USER1).subscribe(new Consumer<String>() {
            @Override
            public void accept(String key) throws Exception {
                Log.d(TAG, "2. Scores of " + key + " changed to " + Preflin.from(SP_SCORES)
                        .getDoubleList(key)
                        .toString());
            }
        });
        disposables.add(disposable2);

        new Handler().postDelayed(new Runnable() {
            @Override
            public void run() {
                Preflin.from(SP_SCORES).putDoubleList(USER1, Arrays.asList(3.3, 4.4, 5.5));
            }
        }, 1000);

        //Check Key exists or not
        Log.d(TAG, "isExist BOOK1=" + Preflin.isKeyExists(BOOK1));

        //Remove element by Key
        Log.d(TAG, "delete BOOK1=" + Preflin.deleteValue(BOOK1));
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();

        if (disposables != null) {
            disposables.dispose();
            disposables = null;
        }
    }

    static class Book implements Serializable {
        String title;
        int page;

        public Book() {
        }

        public Book(String title, int page) {
            this.title = title;
            this.page = page;
        }

        public String getTitle() {
            return title;
        }

        public void setTitle(String title) {
            this.title = title;
        }

        public int getPage() {
            return page;
        }

        public void setPage(int page) {
            this.page = page;
        }

        @Override
        public String toString() {
            return "Book{" + "title='" + title + '\'' + ", page=" + page + '}';
        }
    }
}
